mm2src/common/patterns/state_machine.rs

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    pub async fn run(mut self, initial_state: impl State<Ctx = Ctx, Result = Result>) -> Result {
+        let mut state: Box<dyn State<Ctx = Ctx, Result = Result>> = Box::new(initial_state);
+        loop {
+            let result = state.on_changed(&mut self.ctx).await;
+            let next_state = match result {
+                StateResult::ChangeState(ChangeGuard { next }) => next,
+                StateResult::Finish(ResultGuard { result }) => return result,
+            };
+
+            state = next_state;
+        }
+    }
+}
+
+pub trait TransitionFrom<Prev> {}
+
+/// Prevent implementing [`TransitionFrom<T>`] for `Next` If `T` implements `LastState` already.
+impl<T, Next> !TransitionFrom<T> for Next
+where
+    T: LastState,
+    // this bound is required to prevent conflicting implementation with `impl<T> !TransitionFrom<T> for T`.
+    (T, Next): NotSame,
+{
+}
+
+/// Prevent implementing [`TransitionFrom<T>`] for itself.
+impl<T> !TransitionFrom<T> for T {}
+
+#[async_trait]
+pub trait State: Send + 'static {
+    type Ctx: Send;
+    type Result;
+
+    /// An action is called on entering this state.
+    /// To change the state to another one in the end of processing, use [`StateExt::change_state`].
+    /// For example:
+    /// ```rust
+    /// return Self::change_state(next_state);
+    /// ```
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result>;
+}
+
+pub trait StateExt {
+    /// Change the state to the `next_state`.
+    /// This function performs the compile-time validation whether this state can transition to the `Next` state,
+    /// i.e checks if `Next` implements [`Transition::from(ThisState)`].
+    fn change_state<Next>(next_state: Next) -> StateResult<Next::Ctx, Next::Result>
+    where
+        Self: Sized,
+        Next: State + TransitionFrom<Self>,
+    {
+        StateResult::ChangeState(ChangeGuard::next(next_state))
+    }
+}
+
+impl<T: State> StateExt for T {}
+
+#[async_trait]
+pub trait LastState: Send + 'static {
+    type Ctx: Send;
+    type Result;
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result;
+}
+
+#[async_trait]
+impl<T: LastState> State for T {
+    type Ctx = T::Ctx;
+    type Result = T::Result;
+
+    /// The last state always returns the result of the state machine calculations.
+    async fn on_changed(self: Box<Self>, ctx: &mut T::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        let result = LastState::on_changed(self, ctx).await;
+        StateResult::Finish(ResultGuard::new(result))
+    }
+}
+
+pub enum StateResult<Ctx, Result> {
+    ChangeState(ChangeGuard<Ctx, Result>),
+    Finish(ResultGuard<Result>),
+}
+
+/* vvv The access guards that prevents the user using this pattern from entering an invalid state vvv */
+
+/// An instance of `ChangeGuard` can be initialized within `state_machine` module only.
+pub struct ChangeGuard<Ctx, Result> {
```

Comment:
Could you please provide more details on the `ChangeGuard`? It's unclear for me now how it prevents "from entering an invalid state" :slightly_smiling_face:

Reply:
`ChangeGuard` has to be initialized within the `StateExt::change_state` method **only** because this `StateExt::change_state` method performs the compile-time validation of whether `Self` (for example, `ConnectingState`) can transition to the `Next` state (for example, `OpenState`). 
If a user could initialize `ChangeGuard` outside the `state_machine.rs` file, he might enter an invalid state (for example, `ClosingState -> OpenState`). So it's the reason why `ChangeGuard` or `ResultGuard` can be initialized within the `state_machine.rs` file only (they have private fields and their constructors are private).

Reply:
I forgot to mention that if you want to change the state, you have to return `StateResult::ChangeState(ChangeGuard)`. But you cannot construct `ChangeGuard` manually, so you will do something like `return Self::change_state` where `Self::change_state` performs the transition validation in compile-time and returns `StateResult::ChangeState(ChangeGuard)`.

Reply:
As you noticed yesterday, user implements `State::on_changed` for his custom state. This method should return `StateResult::ChangeState(ChangeGuard)` to change the state to the next one. So user has to use `Self::change_state` that protects the user from entering an invalid state